colorbleed/plugins/maya/publish/extract_look.py

```diff
@@ -2,16 +2,103 @@
 import json
 import tempfile
 import contextlib
+import subprocess
 from collections import OrderedDict
 
 from maya import cmds
 
 import pyblish.api
 import avalon.maya
+from avalon import io
 
 import colorbleed.api
 import colorbleed.maya.lib as lib
 
+# Modes for transfer
+COPY = 1
+HARDLINK = 2
+
+
+def source_hash(filepath, *args):
+    """Generate simple identifier for a source file.
+
+    This is used to identify whether a source file has previously been
+    processe into the pipeline, e.g. a texture.
+
+    The hash is based on source filepath, modification time and file size.
+    This is only used to identify whether a specific source file was already
+    published before from the same location with the same modification date.
+
+    We opt to do it this way as opposed to Avalanch C4 hash as this is much
+    faster and predictable enough for all our production use cases.
+
+    Args:
+        filepath (str): The source file path.
+
+    You can specify additional arguments in the function
+    to allow for specific 'processing' values to be included.
+
+    """
+    # We replace dots with comma because . cannot be a key in a pymongo dict.
+    return "|".join([
+                        os.path.normpath(os.path.abspath(filepath))
+                        .lower().replace("\\", "/"),
+                        str(os.path.getmtime(filepath)),
```

Comment:
I was snooping around, and although this is an old PR which might be out-dated, but since I got this problem a few days ago so I am just leaving a note here anyway :P

The file's modification time might lose a little accuracy after copied by Python, you may see this [post](https://stackoverflow.com/a/17087580/4145300) for detail.

So after a few tests with my site's storage server, I end up using this `math.floor(mtime * 10000) / float(10000)` to chop off the rest digit (floor at 4 decimal places) to get my *soft* mtime.

One other thing is you might want to use `shutil.copy2` to preserve the file's metadata (including mtime) when copying files.

Hopes this helps :D

Reply:
Thanks for dropping those comments - definitely will help when/if we start adopting is. At this stage we never merged this for production as I didn't test it that thoroughly and the need at the time went away as a production got cancelled.

However, and thanks a ton for mentioning this one:

> The file's modification time might lose a little accuracy after copied by Python, you may see this post for detail.

That must be one that is insanely hard to track down if you wouldn't know exactly what was going on. So thanks, that will probably save me time. :D

Reply:
great comment. thanks. We've actually deployed this as it's causing no harm and saves some space. It doesn't quite catch all scenarios, but still makes a difference

Reply:
Are you also using the `maketx` functionality? Or solely the hardlinking?

Reply:
maketx as well. we're only publishing .tx files. we also deal with colourspace transforms upon publishing. Taking them from the file node. Even though that part needs some love.

Reply:
> we also deal with colourspace transforms upon publishing. Taking them from the file node. Even though that part needs some love.

Definitely point me to what you're addressing and how if you have the code available somewhere. :) But thanks for the details!

Reply:
The code is here, but it's very basic at this point. https://github.com/pypeclub/pype/blob/c20f4ba5ee7cc68022a9d0eaae2440a685d74ff3/pype/plugins/maya/publish/extract_look.py#L180